Ensure Onecore synthDriver still loads when the configured voice is missing

michaelDCurran opened the pull request:
### Link to issue number:
Fixes #7553 

### Summary of the issue:
If the Onecore synth driver is set to a particular voice, and then that voice is uninstalled (by the user or Windows itself), the synth driver will fail to load.

### Description of how this pull request fixes the issue:
When the Onecore synth driver is initialized, it now sets a default voice that it knows definitely exists. To choose this voice, it tries to match the user's Windows language as best as possible. Failing this, it just chooses the first valid voice available in the list of voices.
To get the user's Windows language, some code has been split out from languageHandler.setLanguage into a new languageHandler.getWindowsLanguage function.
The refactoring of the voice handling for the Onecore synth driver also means that now changing the voice is much faster than it used to be, as all voices are only fetched once.

### Testing performed:
* Loaded the Onecore synth driver, switched to a particular voice, and saved configuration.
* Manually uninstalled that voice, and started NVDA again. NVDA defaulted back to the first voice for my configured language. Previously it would have failed to load the driver.


### Known issues with pull request:

### Change log entry:
Bug fixes:
* The Windows Onecore Voices synth driver no longer fails to load if the configured voice has been uninstalled.
* Changing voices in the Windows Onecore Voices synth driver is now a lot faster.

michaelDCurran committed:
languageHandler: split out the code that fetches the user's Windows language from setLanguage into a new getWindowsLanguage function so that it can be called separately.

michaelDCurran committed:
Onecore synthDriver: Ensure it does not fail to load if the configured voice has been uninstalled. Also majorly speed up changing voices.
Specifically:
* When fetching the list of voices from Onecore speech, also tet the language for each voice.
* Choose the best default voice on initialization, trying to match the user's Windows language as best as possible. If this fails, just choose the first valid voice it can find.
* Validate and filter the list of available voices as early as possible. This ensures that available voices are not fetched more than once, as the actual call to get the voices from Onecore can sometimes take up to 600 ms or more. thus this was majorly slowing down changing voices.

leonardder commented:
I've tested the OneCore synthesizer on Windows Server 2016, which had no voices installed by that time. Initializing the synth failed with a weird WindowsError. I'd say this was expected, but may be you could check for available in the check function while at it?

michaelDCurran commented:
Checking for availableVoices takes about 600 ms on some machines. Doing 
this in check will be not practical.

feerrenrut commented on source/synthDrivers/oneCore.py:
```diff
@@ -221,24 +223,45 @@ def _callback(self, bytes, len, markers):
+		language=language.replace('-','_')
+		return VoiceInfo(ID,name,language=language)
+
+	def _getAvailableVoices(self):
```
I dont think this will be cached by default, is that a problem? It seems like this might be a slow operation.

michaelDCurran replied:
It is very slow, yes. However, the SynthDriver base class has an availableVoices property that by default calls _getAvailableVoices. This property is cached. As long as we never call _getAvailableVoices we are okay.

leonardder commented on source/languageHandler.py:
```diff
@@ -137,27 +137,34 @@ def pgettext(context, message):
+	"""
+	Fetches the locale name of the user's configured language in Windows.
+	"""
+	windowsLCID=ctypes.windll.kernel32.GetUserDefaultUILanguage()
```
Now we are no longer required to support Windows XP, I'd like to suggest using [GetUserPreferredUILanguages](https://msdn.microsoft.com/en-us/library/windows/desktop/dd318139(v=vs.85).aspx)

michaelDCurran commented:
What do you see as the advantages to using GetUserPreferredUILanguages? 
Are you imagining that for Onecore speech we might try each language 
until we get a voice?
If we were only interested in the first language, I assume there is no 
advantage to this function over UserDefaultUILanguage?

leonardder commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [6 mrt. 2018 10:57 CET](https://github.com/nvaccess/nvda/pull/7999#issuecomment-370726899 "2018-03-06T09:57:26Z - Replied by Github Reply Comments"):
> What do you see as the advantages to using GetUserPreferredUILanguages? 
> If we were only interested in the first language, I assume there is no 
> advantage to this function over UserDefaultUILanguage?

The advantage of using GetUserPreferredUILanguages is that you don't have to do a subsequent call to LCIDToLocaleName. The comments above that call suggest that it is not yet clear what to do by default, either quering the locale id in the locale.windows_locale dictionary, or calling LCIDToLocaleName.

As part of #7949, I switched from GetDateFormat to GetDateFormatEx, because of Microsoft stating their preference of using a locale name over an locale id. From the GetDateFormat page is the following quote:

> Note: For interoperability reasons, the application should prefer the GetDateFormatEx function to GetDateFormat because Microsoft is migrating toward the use of locale names instead of locale identifiers for new locales. Any application that will be run only on Windows Vista and later should use GetDateFormatEx.

Note that I'm only suggesting this as an alternative, I'd be perfectly ok with it if you keep it as it is.

michaelDCurran commented:
Do you know if there is a function to convert an lcid to a localeName? 
It would be nice if we are going to switch to the other function that we 
also allow Windows itself to convert from lcid to locale name, such as 
when receiving the language ID from MS Word, so that the results are the 
same for both functions. Currently we map from lcid to localeName with a 
hard-coded list.


Either way though, I'd be tempted to leave changing the functions at all 
for this particular PR as I'm not totally convinced that Windows 7 
actually contains a complete list of locale names. We'd have to write 
some kind of test for that I guess, even if it was just to prove it once.

leonardder commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [6 Mar 2018, 18:32 CET](https://github.com/nvaccess/nvda/pull/7999#issuecomment-370862268 "2018-03-06T17:32:05Z - Replied by Github Reply Comments"):
> Do you know if there is a function to convert an lcid to a localeName?

Yes, this is [LCIDToLocaleName](https://msdn.microsoft.com/en-us/library/windows/desktop/dd318698%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396) and it is already used in the languageHandler code.

> Currently we map from lcid to localeName with a 
> hard-coded list.

Yes, but when that fails, the code falls back to LCIDToLocaleName.

> Either way though, I'd be tempted to leave changing the functions at all 
> for this particular PR as I'm not totally convinced that Windows 7 
> actually contains a complete list of locale names. We'd have to write 
> some kind of test for that I guess, even if it was just to prove it once.

Hmm yes. But that ought to be tested on a Windows 7 system. I agree, let's keep it as it is for this particular pr.

michaelDCurran commented:
Err... don't mind me. I was commenting in the middle of the night :)

But yes, in future we could switch to using getUserPreferredUILanguages 
and also remove the hard-coded stuff from lcidToLocaleName.

michaelDCurran committed:
Address review actions.

michaelDCurran committed:
Merge branch 'master' into i7553

michaelDCurran commented:
@feerrenrut: review actions have been addressed.

michaelDCurran commented:
Incubating to give people a good chance to test before 2018.2. But @feerrenrut not entirely sure if you were had finished with this. I addressed all your review comments.

feerrenrut reviewed (commented):
I looked over this again, and I am happy with the changes.
